pkg/resource/deploy/step_generator.go

```diff
@@ -1870,40 +1998,68 @@ func (sg *stepGenerator) providerChanged(urn resource.URN, old, new *resource.St
 }
 
 // diff returns a DiffResult for the given resource.
-func (sg *stepGenerator) diff(urn resource.URN, old, new *resource.State, oldInputs, oldOutputs,
-	newInputs resource.PropertyMap, prov plugin.Provider, allowUnknowns bool,
-	ignoreChanges []string,
-) (plugin.DiffResult, error) {
+func (sg *stepGenerator) diff(
```

Comment:
I think we can update this comment to talk about the promise returned, and flesh it out generally.